Added PBN0020 ShouldDeclareDefault analyzer

menees opened the pull request:
I've added PBN0020:DataContractAnalyzer.ShouldDeclareDefault to ensure bool and enum members using
initializers also declare a DefaultValue so they'll be sent correctly (instead of omitted). This analyzer is based on [Marc's SO response](https://stackoverflow.com/a/3162253/1882616) to a question about when bool members weren't being sent/updated as expected. After this burned me a couple of times, I decided to add an analyzer to catch it.

For example, given these members in a ProtoContract:
``` C#
[ProtoMember(1)]
public bool Recursive { get; set; } = true;

[ProtoMember(2)]
public DayOfWeek Day { get; set; } = DayOfWeek.Monday;
```

This analyzer will produce the following two warnings:
* PBN0020: Field 'Recursive' should use [DefaultValue(true)] to ensure its value is sent since it's initialized to a non-default value.
* PBN0020: Field 'Day' should use [DefaultValue(DayOfWeek.Monday)] to ensure its value is sent since it's initialized to a non-default value.

I didn't add an automated code-fixer because it's a lot more work, and there are several supported ways to suppress these warnings based on Marc's SO answer. You can use the DefaultValue attribute, you can use ProtoMember's IsRequired=true property, or you can add a bool ShouldSerialize{MemberName}() method. The warning text recommends using the DefaultValue attribute like:

``` C#
[ProtoMember(1)]
[DefaultValue(true)]
public bool Recursive { get; set; } = true;

[ProtoMember(2)]
[DefaultValue(DayOfWeek.Monday)]
public DayOfWeek Day { get; set; } = DayOfWeek.Monday;
```

This analyzer only looks for simple initializers using an "= value" format for bool (= true) and enum (= non-zero field). So it can't detect members initialized to non-default values by constructors or by more complex expressions (e.g., = Convert.ToBoolean(1)). But this simple logic handles the cases I've encountered, and I'm hoping it can help others.

I also added this analyzer because I needed my own protobuf-net.BuildTools package anyway to get the standard analyzer support for C#9 record types that was added in [protobuf-net #783](https://github.com/protobuf-net/protobuf-net/issues/783) since that hasn't been pushed to NuGet.org yet.

mgravell commented:
This looks awesome, thanks hugely; looking now.

mgravell commented:
On the surface this looks great; I'll try to have a play with it tomorrow, but; +1

For the paper-trail etc; can I confirm that you're happy and able to freely contribute this code in line with the library's license, etc.

mgravell commented:
Minor question: do we currently handle integers? floating point? etc; is it worth extending this PR to add those, or should I add those later?

menees commented:
That's an excellent question about ints, floats, etc. Let me handle them by extending this PR. I didn't think about them because I hadn't run into a problem with them (that I'm aware of). But it seems like this analyzer should check them too.

menees commented:
I added support for char and numeric fields. They were a little tricky because C# allows so many different literal escape sequences or suffixes and leading unary operators. I tried to handle all the variations I could see from the C# docs: [char](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/char#literals), [integral](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types#integer-literals), [floating-point](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/floating-point-numeric-types#real-literals).

Also, in good news (sort of), I ran this on my work codebase and found 11 places where someone was assigning a 1 or -1 via an initializer. Now I've added DefaultValue attributes so they'll serialize correctly if a 0 is ever assigned to them. Good catch!

menees commented:
> For the paper-trail etc; can I confirm that you're happy and able to freely contribute this code in line with the library's license, etc.

Yes, I'm happy and able to freely contribute this code in line with the library's license. Thanks!

mgravell reviewed (approved):
This looks great, thanks
